mayastor/src/bdev/nexus/nexus_child.rs

```diff
@@ -252,65 +259,112 @@ impl NexusChild {
             return Err(ChildError::ChildReadOnly {});
         }
 
-        let bdev = self.bdev.as_ref();
-        let desc = self.bdev_handle.as_ref();
-
-        if bdev.is_none() || desc.is_none() {
-            return Err(ChildError::ChildInvalid {});
-        }
+        let bdev = match self.bdev.as_ref() {
+            Some(dev) => dev,
+            None => {
+                return Err(ChildError::ChildInvalid {});
+            }
+        };
 
-        let bdev = bdev.unwrap();
-        let desc = desc.unwrap();
+        let desc = match self.bdev_handle.as_ref() {
+            Some(handle) => handle,
+            None => {
+                return Err(ChildError::ChildInvalid {});
+            }
+        };
 
         let block_size = bdev.block_len();
-
-        let primary = u64::from(block_size);
-        let secondary = bdev.num_blocks() - 1;
-
         let mut buf = desc
             .dma_malloc(block_size as usize)
             .context(LabelAlloc {})?;
 
-        self.read_at(primary, &mut buf)
+        self.read_at(0, &mut buf).await.context(LabelRead {})?;
+        let mbr = match Pmbr::from_slice(&buf.as_slice()[440 .. 512]) {
```

Comment:
This makes an invalid MBR an error, I don't think that is needed. The PMBR and the GPT label do not depend on one other AFAIK, right?

Reply:
There is a dependency between the MBR and the GPT. Acording to the spec, you should not attempt to reconstruct a primary GPT from the secondary unless it is consistent with the MBR. This is to avoid the following scenario:
A third party reformats the disk as MBR rather than GPT. This will probably result in the primary GPT label being overwritten more or less immediately, but more than likely leave the secondary GPT label situated at the end of the disk completeley intact - even though it is now invalid. If we reconstruct the primary GPT from this secondary without first checking for the existence of a protective partition in the MBR, then we will not detect that we are potentitally overwriting data.

Reply:
Hi, @cjones1024 It been a while since I've read the spec,  Based on your explanation what it would amount to is that if the first 512 bytes are toast, we simply discard the whole partitioning scheme. This seems rather harsh. 

Also, we know that we only care about GPT and so we don't do anything with the PMBR whenever its good or bad. All it would say in our case is that it contains one partition that spans the whole disk which is lie -- and we lie about it because the MBR code knows nothing about GPT and we mostly use it to shut up boot code anyway. 

The only reason I added it the first place was because gptdisk said that the PMBR was invalid/missing but that was all it really did: noise. 

Anyways, if the spec says its invalid, let us follow the spec then, but please add a comment for future reference, thanks.

Reply:
Maybe a compromise is to just issue a warning but not consider this to be a fatal error?